[case mypy_ini_exclude_simple]
# Copied from testCmdlineExclude
[file lala/apkg.py]
1()
[file b/bpkg.py]
1()  # E: "int" not callable
[file c/cpkg.py]
1()  # E: "int" not callable
[file mypy.ini]
\[mypy]
exclude = lala

[case mypy_toml_exclude_simple]
[file lala/apkg.py]
1()
[file b/bpkg.py]
1()  # E: "int" not callable
[file c/cpkg.py]
1()  # E: "int" not callable
[file pyproject.toml]
\[tool.mypy]
exclude = "lala"

[case mypy_toml_exclude_list]
[file lala/apkg.py]
1()
[file b/bpkg.py]
1()
[file c/cpkg.py]
1()  # E: "int" not callable
[file pyproject.toml]
\[tool.mypy]
exclude = [
    "lala",
    "b(/|\\\\).*pkg",
]

[case mypy_ini_exclude_multi_line_verbose_regex]
[file lala/apkg.py]
1()
[file b/bpkg.py]
1()
[file c/cpkg.py]
1()  # E: "int" not callable
[file mypy.ini]
\[mypy]
exclude = (?x)(
    ^apkg\.py$    # files named "one.py"
    | ^b/   # or files starting with "three."
  )

[case mypy_ini_exclude_multi_line_verbose_regex_not_working]
[file lala/apkg.py]
1()  # E: "int" not callable
[file b/bpkg.py]
1()  # E: "int" not callable
[file c/cpkg.py]
1()  # E: "int" not callable
[file mypy.ini]
\[mypy]
exclude = (
    ^apkg\.py$    # files named "one.py"
    | ^b/   # or files starting with "three."
  )

[case mypy_toml_exclude_with_overrides_inheritance]
[file lala/a/apkg.py]
1()
[file lala/a/other.py]
1()  # E: "int" not callable
[file lala/a/inherited_exclude.py]
1()
[file lala/a/bpkg.py]
1()
[file lala/b/cpkg.py]
1()  # E: "int" not callable
[file pyproject.toml]
\[tool.mypy]
exclude = "lala(/|\\\\)a.*pkg"
\[[tool.mypy.overrides]]
module = 'lala.a.*'
exclude = "inherited_exclude"

[case mypy_ini_excluded_file_can_have_syntax_errors]
[file m.py]
1[]
[file n.py]
1[]  # E: Invalid syntax
[file mypy.ini]
\[mypy]
exclude = 'm.py$'

[case mypy_setup_ini_with_multiline_string]
# See also https://github.com/zonyitoo/rust-ini/issues/138
def foo(x: int = None) -> None: ...
foo(1)
foo(None)

[file mypy.ini]
[section1]
a =
  b
  c
[mypy]
implicit_optional = True

[case invalid_directives]
# mypy: always-false="FOO,BAR";,
# mypy: always-false="FOO,BAR""bla",
# mypy: "always-true=foo"
[out]
__main__:1: error: Content after quote in configuration comment
__main__:3: error: Content after quote in configuration comment
__main__:3: error: Quotes should not be part of the key
__main__:3: error: Unrecognized option: "always_true = oo"

[case untyped_defs_correct_errors]
[file checked.py]
def x(y):  # E: Function is missing a type annotation  [no-untyped-def]
    1()  # E: "int" not callable  [operator]

def y(z: int): ...  # E: Function is missing a return type annotation  [no-untyped-def]

[file unchecked.py]
def x(y):
    1()

def y(z: int): ...  # E: Function is missing a return type annotation  [no-untyped-def]

# mypy: allow-untyped-defs, no-check-untyped-defs

[file mypy.ini]
[mypy]
check_untyped_defs = true
disallow_untyped_defs = true
disallow_incomplete_defs = true
show_error_codes = true

[mypy-unchecked]
no_implicit_reexport = true

[case explicit_mode_with_mypy_config_should_not_lead_to_mypy_mode]
# flags: --mode default

def f():
    x: int = ""  # E: Incompatible types in assignment (expression has type "str", variable has type "int")

[file pyproject.toml]
[tool.mypy]

[case auto_mode_should_use_mypy_in_case_of_mypy_config]
# flags: --auto-mode

def f():
    x: int = ""  # N: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs

[file pyproject.toml]
[tool.mypy]

[case auto_mode_with_specified_mode_in_config]
# flags: --auto-mode

def f():
    x: int = ""  # N: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs

[file pyproject.toml]
[tool.zuban]
mode = "mypy"

[case auto_mode_with_explicit_default_mode_and_mypy_config_should_use_explicit_mode]
# flags: --auto-mode

def f():
    x: int = ""  # E: Incompatible types in assignment (expression has type "str", variable has type "int")

[file pyproject.toml]
[tool.zuban]
mode = "default"

[tool.mypy]

[case pyproject_ignore_gitignore_files]
# flags: --exclude-gitignore
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()  # E: "int" not callable
[file c.py]
1()
[file d.py]
1()
[file dir/c.py]
1()  # E: "int" not callable
[file dir/d.py]
1()

[file .gitignore]
/c.py
d.py

[tool.zuban]
exclude = "a.py"

[case pyproject_ignore_nested_gitignore_files]
# flags: --exclude-gitignore
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()  # E: "int" not callable
[file c.py]
1()
[file dir/a.py]
1()
[file dir/nested/a.py]
1()  # E: "int" not callable
[file dir/b.py]
1()
[file dir/c.py]
1()  # E: "int" not callable

[file .gitignore]
/c.py

[file dir/.gitignore]
/a.py
b.py

[case pyproject_gitignore_whitelist]
# flags: --exclude-gitignore
[file a.py]
1()
[file b.py]
1()  # E: "int" not callable

[file .gitignore]
*.py
!b.py

[case pyproject_gitignore_invalidation]
# flags: --exclude-gitignore
[file a.py]
1()
[file b.py]
1()
[file .gitignore]
b.py
[delete .gitignore.2]
[file .gitignore.3]
a.py
[file .gitignore.4]
b.py
[out]
a.py:1: error: "int" not callable
==
a.py:1: error: "int" not callable
b.py:1: error: "int" not callable
==
b.py:1: error: "int" not callable
==
a.py:1: error: "int" not callable

[case pyproject_gitignore_flag_default_in_mypy_mode]
# flags: --mode mypy
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()  # E: "int" not callable
[file .gitignore]
b.py

[case pyproject_gitignore_flag_default_in_default_mode]
# flags: --mode default
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()
[file .gitignore]
b.py

[case pyproject_no_exclude_gitignore_flag]
# flags: --no-exclude-gitignore
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()  # E: "int" not callable
[file .gitignore]
b.py

[case pyproject_exclude_gitignore_flag]
# flags: --exclude-gitignore
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()
[file .gitignore]
b.py

[case pyproject_exclude_gitignore_in_config]
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()
[file .gitignore]
b.py
[file pyproject.toml]
[tool.mypy]
exclude_gitignore = true

[case pyproject_no_exclude_gitignore_in_config]
[file a.py]
1()  # E: "int" not callable
[file b.py]
1()  # E: "int" not callable
[file .gitignore]
b.py
[file pyproject.toml]
[tool.zuban]
exclude_gitignore = false
